Improve Active Job test helpers

bogdanvlviv opened the pull request:
- ec2e8f645eded183224420e209420376c63b99bb

    Allow `:queue` option to `perform_enqueued_jobs`.

    If the `:queue` option is specified, then only the job(s) enqueued to
    a specific queue will be performed.

    Example:
    ```
    def test_perform_enqueued_jobs_with_queue
      perform_enqueued_jobs queue: :some_queue do
        MyJob.set(queue: :some_queue).perform_later(1, 2, 3) # will be performed
        HelloJob.set(queue: :other_queue).perform_later(1, 2, 3) # will not be performed
      end
      assert_performed_jobs 1
    end
    ```

    Follow up #33265

    [bogdanvlviv & Jeremy Daer]

<hr/>

- e0cf042fa2c499c2e0b8fe95ec02ed79f9178feb

    Fix `perform_enqueued_jobs`

    Set
    ````ruby
    queue_adapter.perform_enqueued_jobs = true
    queue_adapter.perform_enqueued_at_jobs = true
    queue_adapter.filter = only
    queue_adapter.reject = except
    queue_adapter.queue = queue
    ```
    if block given.
    Execution of `flush_enqueued_jobs` doesn't require that.

<hr/>

- d50fb21e4d7a526de3b0edb9bb5032d4be6175ae

    Allow `:queue` option to `assert_performed_jobs`.

    If the `:queue` option is specified, then only the job(s) enqueued to a specific
    queue will be performed.

    Example:
    ```ruby
    def test_assert_performed_jobs_with_queue_option
      assert_performed_jobs 1, queue: :some_queue do
        HelloJob.set(queue: :some_queue).perform_later("jeremy")
        HelloJob.set(queue: :other_queue).perform_later("bogdan")
      end
    end
    ```

<hr/>

- de4420da44ecce68a6b7607b828c69959a5f738e

    Allow `:queue` option to `assert_no_performed_jobs`.

    If the `:queue` option is specified, then only the job(s) enqueued to a specific
    queue will not be performed.

    Example:
    ```ruby
    def test_assert_no_performed_jobs_with_queue_option
      assert_no_performed_jobs queue: :some_queue do
        HelloJob.set(queue: :other_queue).perform_later("jeremy")
      end
    end
    ```

<hr/>

- 2bf8b4eb0eceb37fc278c71371a659979844dfb2

    Add changelog entry about adding `:queue` option to job assertions and helpers

    Note that it removes changelog entry of #33265 since the entry in this commits
    includes that too.

<hr/>

- 11634e8ef85870535bad61b999e0f9ec36f6f963

    Fix `assert_performed_jobs` and `assert_no_performed_jobs`

    Execution of `assert_performed_jobs`, and `assert_no_performed_jobs`
    without a block should respect passed `:except`, `:only`, and `:queue` options.

<hr/>

- 2ec60fb8186ad3b227341b7d1874e3f13bd34279

    Allow `assert_performed_with` to be called without a block.

    Example:
    ```ruby
    def test_assert_performed_with
      MyJob.perform_later(1,2,3)

      perform_enqueued_jobs

      assert_performed_with(job: MyJob, args: [1,2,3], queue: 'high')
    end
    ```

    Follow up #33626.

<hr/>

- b7beb5d4e5a5f642d172002723ef269d0c6a0bfd

    Fix formatting of `ActiveJob::TestHelper` api docs

<hr/>

- b8576425eeee2eaa82c5acb08227f5bb2dcd07b7

    DRY in `assert_enqueued_jobs`

<hr/>

Note that these changes keep backward compatibility, so It shouldn't break users' tests during an upgrade to 6.0.

bogdanvlviv committed:
Allow `:queue` option to `perform_enqueued_jobs`.

If the `:queue` option is specified, then only the job(s) enqueued to
a specific queue will be performed.

Example:
```
def test_perform_enqueued_jobs_with_queue
  perform_enqueued_jobs queue: :some_queue do
    MyJob.set(queue: :some_queue).perform_later(1, 2, 3) # will be performed
    HelloJob.set(queue: :other_queue).perform_later(1, 2, 3) # will not be performed
  end
  assert_performed_jobs 1
end
```

Follow up #33265

[bogdanvlviv & Jeremy Daer]

bogdanvlviv committed:
Fix `perform_enqueued_jobs`

Set
````
queue_adapter.perform_enqueued_jobs = true
queue_adapter.perform_enqueued_at_jobs = true
queue_adapter.filter = only
queue_adapter.reject = except
queue_adapter.queue = queue
```
if block given.
Execution of `flush_enqueued_jobs` doesn't require that.

bogdanvlviv committed:
Allow `:queue` option to `assert_performed_jobs`.

If the `:queue` option is specified, then only the job(s) enqueued to a specific
queue will be performed.

Example:
```
def test_assert_performed_jobs_with_queue_option
  assert_performed_jobs 1, queue: :some_queue do
    HelloJob.set(queue: :some_queue).perform_later("jeremy")
    HelloJob.set(queue: :other_queue).perform_later("bogdan")
  end
end
```

bogdanvlviv committed:
Allow `:queue` option to `assert_no_performed_jobs`.

If the `:queue` option is specified, then only the job(s) enqueued to a specific
queue will not be performed.

Example:
```
def test_assert_no_performed_jobs_with_queue_option
  assert_no_performed_jobs queue: :some_queue do
    HelloJob.set(queue: :other_queue).perform_later("jeremy")
  end
end
```

bogdanvlviv committed:
Fix `assert_performed_jobs` and `assert_no_performed_jobs`

Execution of `assert_performed_jobs`, and `assert_no_performed_jobs`
without a block should respect passed `:except`, `:only`, and `:queue` options.

bogdanvlviv committed:
Allow `assert_performed_with` to be called without a block.

Example:
```
def test_assert_performed_with
  MyJob.perform_later(1,2,3)

  perform_enqueued_jobs

  assert_performed_with(job: MyJob, args: [1,2,3], queue: 'high')
end
```

Follow up #33626.

bogdanvlviv committed:
Fix formatting of `ActiveJob::TestHelper` api docs

bogdanvlviv committed:
DRY in `assert_enqueued_jobs`

jeremy commented on activejob/CHANGELOG.md:
````diff
@@ -1,3 +1,57 @@
+    ```
+
+    *bogdanvlviv*
+
````
Let's combine into one changelog entry that heralds these together, that `:queue` may be used with job assertions & helpers.

bogdanvlviv replied:
Done in https://github.com/rails/rails/commit/c6f4ddf34531f5bc017eb7e5cde0fabd5dd411cc.

jeremy commented on activejob/CHANGELOG.md:
```diff
@@ -11,7 +65,7 @@
     *Zach Kemp*
 
-*   Allow `queue` option to `assert_no_enqueued_jobs`.
+*   Allow `:queue` option to `assert_no_enqueued_jobs`.
```
👍

jeremy commented on activejob/lib/active_job/queue_adapters/test_adapter.rb:
```diff
@@ -55,9 +55,11 @@ def perform_or_enqueue(perform, job, job_data)
+            job.queue_name != queue.to_s
           else
             false
           end
```
Can hoist up the `queue` conditional since it applies to every clause here:
```ruby
def filtered?(job)
  filtered_queue?(job) || filtered_job_class?(job)
end

def filtered_queue?(job)
  if queue
    job.queue_name != queue.to_s
  end
end

def filtered_job_class?(job)
  if filter
    !Array(filter).include?(job.class)
  elsif reject
    Array(reject).include?(job.class)
  end
end
```

bogdanvlviv replied:
:heart_eyes: Thanks for the tip. I like this simplified format. I applied this suggestion in https://github.com/rails/rails/commit/ec2e8f645eded183224420e209420376c63b99bb and added you to authors of this commit.

jeremy commented on activejob/lib/active_job/test_helper.rb:
```diff
@@ -52,7 +52,7 @@ def after_teardown # :nodoc:
     end
 
-    # Specifies the queue adapter to use with all active job test helpers.
+    # Specifies the queue adapter to use with all Active Job test helpers.
```
💅  love these edits

MarcSteven replied:
cool edit

jeremy commented on activejob/lib/active_job/test_helper.rb:
```diff
@@ -118,13 +118,16 @@ def queue_adapter_for_test
         actual_count = enqueued_jobs_with(only: only, except: except, queue: queue)
-        assert_equal number, actual_count, "#{number} jobs expected, but #{actual_count} were enqueued"
       end
+      assert_equal number, actual_count, "#{number} jobs expected, but #{actual_count} were enqueued"
```
Nice style change. To confirm, I'm not seeing any implementation change here?

jeremy commented on activejob/lib/active_job/test_helper.rb:
```diff
@@ -176,7 +179,7 @@ def assert_no_enqueued_jobs(only: nil, except: nil, queue: nil, &block)
     # Asserts that the number of performed jobs matches the given number.
     # If no block is passed, <tt>perform_enqueued_jobs</tt>
-    # must be called around the job call.
+    # must be called around or after the job call.
```
👍

jeremy commented on activejob/lib/active_job/test_helper.rb:
```diff
@@ -186,10 +189,11 @@ def assert_no_enqueued_jobs(only: nil, except: nil, queue: nil, &block)
+    #
+    #     perform_enqueued_jobs
+    #
+    #     assert_performed_jobs 2
```
We're showing two API calling styles here. Let's provide some guidance about when you'd use one vs. the other.

bogdanvlviv replied:
Hm. Then we should provide guidance about the one vs the other to every helper/assertion.
I believe that this difference is explicit from the current explanations like
`"If a block is passed, that block performs all of the jobs that were enqueued throughout the duration of the block and asserts that the job has been performed with the given arguments in the block."` and examples.
Let's go with as it is for now?

jeremy replied:
Agreed 👍

jeremy reviewed (approved):
Fantastic work @bogdanvlviv.

jeremy commented:
@bogdanvlviv Do you think 2b471e1 is safe to backport to 5-2-stable without breaking compatibility?

bogdanvlviv committed:
Add changelog entry about adding `:queue` option to job assertions and helpers

Note that it removes changelog entry of #33265 since the entry in this commits
includes that too.

bogdanvlviv commented:
>  @bogdanvlviv Do you think 2b471e1 is safe to backport to 5-2-stable without breaking compatibility? (As far as I can see, it's a bugfix.)


@jeremy We shouldn't backport b43d3c07b5eeff5c0569b901933d17071e387b36 to `5-2-stable` since execution of `perform_enqueued_jobs` without a block was allowed only on the `master` branch in #33626.

The current implementation of `perform_enqueued_jobs` on `5-2-stable`: https://github.com/rails/rails/blob/5-2-stable/activejob/lib/active_job/test_helper.rb#L368L368-L386

kamipo commented:
How about an alternative of b43d3c07b5eeff5c0569b901933d17071e387b36 to touch few code like here?

```diff
diff --git a/activejob/lib/active_job/test_helper.rb b/activejob/lib/active_job/test_helper.rb
index b45cc57fb0..b130b9604b 100644
--- a/activejob/lib/active_job/test_helper.rb
+++ b/activejob/lib/active_job/test_helper.rb
@@ -404,6 +404,8 @@ def assert_performed_with(job: nil, args: nil, at: nil, queue: nil)
     #   end
     #
     def perform_enqueued_jobs(only: nil, except: nil)
+      return flush_enqueued_jobs(only: only, except: except) unless block_given?
+
       validate_option(only: only, except: except)
       old_perform_enqueued_jobs = queue_adapter.perform_enqueued_jobs
       old_perform_enqueued_at_jobs = queue_adapter.perform_enqueued_at_jobs
@@ -416,7 +418,7 @@ def perform_enqueued_jobs(only: nil, except: nil)
         queue_adapter.filter = only
         queue_adapter.reject = except
 
-        block_given? ? yield : flush_enqueued_jobs(only: only, except: except)
+        yield
       ensure
         queue_adapter.perform_enqueued_jobs = old_perform_enqueued_jobs
         queue_adapter.perform_enqueued_at_jobs = old_perform_enqueued_at_jobs
```

bogdanvlviv commented:
I want to move 
```
old_perform_enqueued_jobs = queue_adapter.perform_enqueued_jobs
old_perform_enqueued_at_jobs = queue_adapter.perform_enqueued_at_jobs
old_filter = queue_adapter.filter
old_reject = queue_adapter.reject
old_queue = queue_adapter.queue
```
to `begin ... end` too.

Also we should execute `flush_enqueued_jobs` after `validate_option`.

Do you want me to redo b43d3c07b5eeff5c0569b901933d17071e387b36?

kamipo commented:
`enqueued_jobs_with` in `flush_enqueued_jobs` also does `validate_option` in the first line.

https://github.com/rails/rails/blob/265eeb25592cccb1918641290c1b0259f700d02c/activejob/lib/active_job/test_helper.rb#L446-L447

Is there any reason to do `validate_option` twice?

Personally I feel the remaining part in the commit is like a cosmetic (moving a code chunk, changing indentation, for the purpose of the commit).

bogdanvlviv commented:
> https://github.com/rails/rails/pull/33635#issuecomment-414256295

@kamipo Good point. :+1: 

Fixed in e0cf042fa2c499c2e0b8fe95ec02ed79f9178feb.

jeremy reviewed (approved):
👏

Edouard-chin commented:
Those are really great addition to AJ test helper, thanks ❤️ !

bogdanvlviv commented:
Thanks all :bowing_man: !
